From 5a7786cd9077a16c6dc4656f7c3f93f74ad2e928 Mon Sep 17 00:00:00 2001
From: Shwetha G K <quic_kshwetha@quicinc.com>
Date: Wed, 7 Aug 2024 12:10:51 +0530
Subject: [PATCH] cfg80211: Add support to set per link txpower

Fetch and pass the link id to rdev_set_tx_power inorder
to support the per link tx power setting

Change-Id: I642602c619d2541c2dcbe9b3f12e18f3a61bed3f
---
 include/net/cfg80211.h     |  3 ++-
 net/wireless/nl80211.c     |  3 ++-
 net/wireless/rdev-ops.h    |  7 ++++---
 net/wireless/trace.h       | 11 +++++++----
 net/wireless/wext-compat.c |  2 +-
 5 files changed, 16 insertions(+), 10 deletions(-)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index c1e21dc3dcb0..03ecc348cec6 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -4018,6 +4018,7 @@ struct mgmt_frame_regs {
 };
 
 #define CFG80211_PER_LINK_TX_POWER 1
+#define CFG80211_SET_PER_LINK_TX_POWER 1
 
 /**
  * struct cfg80211_ops - backend description for wireless configuration
@@ -4565,7 +4566,7 @@ struct cfg80211_ops {
 	int	(*set_wiphy_params)(struct wiphy *wiphy, u32 changed);
 
 	int	(*set_tx_power)(struct wiphy *wiphy, struct wireless_dev *wdev,
-				enum nl80211_tx_power_setting type, int mbm);
+				enum nl80211_tx_power_setting type, int mbm, int link_id);
 	int	(*get_tx_power)(struct wiphy *wiphy, struct wireless_dev *wdev,
 				unsigned int link_id, int *dbm);
 
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 922b6adcedd1..1908e0aba8c3 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -3576,6 +3576,7 @@ static int nl80211_set_wiphy(struct sk_buff *skb, struct genl_info *info)
 		struct wireless_dev *txp_wdev = wdev;
 		enum nl80211_tx_power_setting type;
 		int idx, mbm = 0;
+		int link_id = nl80211_link_id_or_invalid(info->attrs);
 
 		if (!(rdev->wiphy.features & NL80211_FEATURE_VIF_TXPOWER))
 			txp_wdev = NULL;
@@ -3599,7 +3600,7 @@ static int nl80211_set_wiphy(struct sk_buff *skb, struct genl_info *info)
 			mbm = nla_get_u32(info->attrs[idx]);
 		}
 
-		result = rdev_set_tx_power(rdev, txp_wdev, type, mbm);
+		result = rdev_set_tx_power(rdev, txp_wdev, type, mbm, link_id);
 		if (result)
 			goto out;
 	}
diff --git a/net/wireless/rdev-ops.h b/net/wireless/rdev-ops.h
index e6c4128c89fd..f0625342ec40 100644
--- a/net/wireless/rdev-ops.h
+++ b/net/wireless/rdev-ops.h
@@ -587,11 +587,12 @@ rdev_set_wiphy_params(struct cfg80211_registered_device *rdev, u32 changed)
 
 static inline int rdev_set_tx_power(struct cfg80211_registered_device *rdev,
 				    struct wireless_dev *wdev,
-				    enum nl80211_tx_power_setting type, int mbm)
+				    enum nl80211_tx_power_setting type, int mbm,
+				    int link_id)
 {
 	int ret;
-	trace_rdev_set_tx_power(&rdev->wiphy, wdev, type, mbm);
-	ret = rdev->ops->set_tx_power(&rdev->wiphy, wdev, type, mbm);
+	trace_rdev_set_tx_power(&rdev->wiphy, wdev, type, mbm, link_id);
+	ret = rdev->ops->set_tx_power(&rdev->wiphy, wdev, type, mbm, link_id);
 	trace_rdev_return_int(&rdev->wiphy, ret);
 	return ret;
 }
diff --git a/net/wireless/trace.h b/net/wireless/trace.h
index 118feed517c4..8293505ddf64 100644
--- a/net/wireless/trace.h
+++ b/net/wireless/trace.h
@@ -1598,22 +1598,25 @@ DEFINE_EVENT(wiphy_wdev_evt, rdev_get_tx_power,
 
 TRACE_EVENT(rdev_set_tx_power,
 	TP_PROTO(struct wiphy *wiphy, struct wireless_dev *wdev,
-		 enum nl80211_tx_power_setting type, int mbm),
-	TP_ARGS(wiphy, wdev, type, mbm),
+		 enum nl80211_tx_power_setting type, int mbm, int link_id),
+	TP_ARGS(wiphy, wdev, type, mbm, link_id),
 	TP_STRUCT__entry(
 		WIPHY_ENTRY
 		WDEV_ENTRY
 		__field(enum nl80211_tx_power_setting, type)
 		__field(int, mbm)
+		__field(int, link_id)
 	),
 	TP_fast_assign(
 		WIPHY_ASSIGN;
 		WDEV_ASSIGN;
 		__entry->type = type;
 		__entry->mbm = mbm;
+		__entry->link_id = link_id;
 	),
-	TP_printk(WIPHY_PR_FMT ", " WDEV_PR_FMT ", type: %u, mbm: %d",
-		  WIPHY_PR_ARG, WDEV_PR_ARG,__entry->type, __entry->mbm)
+	TP_printk(WIPHY_PR_FMT ", " WDEV_PR_FMT ", type: %u, mbm: %d link: %d",
+		  WIPHY_PR_ARG, WDEV_PR_ARG,__entry->type, __entry->mbm,
+		  __entry->link_id)
 );
 
 TRACE_EVENT(rdev_return_int_int,
diff --git a/net/wireless/wext-compat.c b/net/wireless/wext-compat.c
index 6fa5fa1924f2..5290ed0d7889 100644
--- a/net/wireless/wext-compat.c
+++ b/net/wireless/wext-compat.c
@@ -973,7 +973,7 @@ static int cfg80211_wext_siwtxpower(struct net_device *dev,
 	}
 
 	wiphy_lock(&rdev->wiphy);
-	ret = rdev_set_tx_power(rdev, wdev, type, DBM_TO_MBM(dbm));
+	ret = rdev_set_tx_power(rdev, wdev, type, DBM_TO_MBM(dbm), -1);
 	wiphy_unlock(&rdev->wiphy);
 
 	return ret;
-- 
2.34.1

